src/components/TVChartContainer/TVChartContainer.tsx

```diff
@@ -0,0 +1,142 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import { TV_SAVE_LOAD_CHARTS } from "config/localStorage";
+import { useLocalStorage, useMedia } from "react-use";
+import { defaultChartProps, disabledFeaturesOnMonile, SaveLoadAdapter } from "./constants";
+import useTVDatafeed from "domain/tradingview/useTVDatafeed";
+import { IChartingLibraryWidget, IPositionLineAdapter } from "./charting_library/charting_library";
+import { getPeriodFromResolutions, supportedResolutions } from "domain/tradingview/helper";
+
+type Props = {
+  symbol: string;
+  chainId: number;
+  savedShouldShowPositionLines: boolean;
+  currentPositions: any[];
+  currentOrders: any[];
+  onSelectToken: () => void;
+  period: string;
+  setPeriod: (period: string) => void;
+};
+
+export default function TVChartContainer({
+  symbol,
+  chainId,
+  savedShouldShowPositionLines,
+  currentPositions,
+  currentOrders,
+  onSelectToken,
+  period,
+  setPeriod,
+}: Props) {
+  const chartContainerRef = useRef<HTMLDivElement | null>(null);
+  const tvWidgetRef = useRef<IChartingLibraryWidget | null>(null);
+  const [chartReady, setChartReady] = useState(false);
+  let [tvCharts, setTvCharts] = useLocalStorage(TV_SAVE_LOAD_CHARTS, []);
+  const datafeed = useTVDatafeed();
+  const isMobile = useMedia("(max-width: 550px)");
+
+  const drawLineOnChart = useCallback(
+    (title, price) => {
+      if (chartReady && tvWidgetRef.current?.activeChart?.().dataReady()) {
+        return tvWidgetRef.current
+          .activeChart()
+          .createPositionLine({ disableUndo: true })
+          .setText(title)
+          .setPrice(price)
+          .setQuantity("")
+          .setLineStyle(1)
+          .setLineLength(1)
+          .setBodyFont(`normal 12pt "Relative", sans-serif`)
+          .setBodyTextColor("#fff")
+          .setLineColor("#3a3e5e")
+          .setBodyBackgroundColor("#3a3e5e")
+          .setBodyBorderColor("#3a3e5e");
+      }
+    },
+    [chartReady]
+  );
+
+  useEffect(() => {
+    const lines: (IPositionLineAdapter | undefined)[] = [];
+    if (savedShouldShowPositionLines) {
+      currentPositions.forEach((position) => {
+        const { open, liq } = position;
+        lines.push(drawLineOnChart(open.title, open.price));
+        lines.push(drawLineOnChart(liq.title, liq.price));
+      });
+      currentOrders.forEach((order) => {
+        lines.push(drawLineOnChart(order.title, order.price));
+      });
+    }
+    return () => {
+      lines.forEach((line) => line?.remove());
+    };
+  }, [currentPositions, savedShouldShowPositionLines, currentOrders, drawLineOnChart]);
+
+  useEffect(() => {
+    if (chartReady && tvWidgetRef.current && symbol !== tvWidgetRef.current?.activeChart?.().symbol()) {
+      tvWidgetRef.current.setSymbol(symbol, tvWidgetRef.current.activeChart().resolution(), () => {});
+    }
+  }, [symbol, chartReady, period]);
```

Comment:
`tvWidgetRef.current` is not part of hook dependencies. 

in case hook was trigger with tvWidgetRef.current === undefined it would never be called again until symbol/chartReady/period are updated

how does that work? maybe there is no need to check `tvWidgetRef.current`?

Reply:
at first glance if `chartReady` is true then tvWidgetRef is defined? and btw why ref is preferable here instead of doing something like `setState(tvWidget)`?

Reply:
I thought of using the state to store the `tvWidget` but I was facing couple of issue.

- I feel using `ref` is a better choice because we are storing the instance of the widget once created and the instance can be manipulated. TradingView charting library is an external system and it makes sense to store it in ref.
- When I was testing to store the widget instance into a state some of the overrides were not working properly. Like one below. With ref it works normally but with state the color does't apply.

```js
 tvWidget?.applyOverrides({
        "paneProperties.background": "#16182e",
        "paneProperties.backgroundType": "solid",
  });
```

Reply:
> I feel using ref is a better choice because we are storing the instance of the widget once created and the instance can be manipulated. TradingView charting library is an external system and it makes sense to store it in ref.

not sure i understand this. currently in production we use setState to store a reference to TV, why is it different here?

> When I was testing to store the widget instance into a state some of the overrides were not working properly. Like one below. With ref it works normally but with state the color does't apply.

i guess that should not depend on whether it's a setState or useRef. if `tvWidget` has a reference to correct object then it should work? wondering how can setState affect it?

> Because we are storing the widget instance inside a ref which can be mutable we can't use them in dependency array.

i believe react checks reference, not its values recursively? 

in general it might not be critical, just trying to understand why ref is better here

Reply:
and for this comment https://github.com/gmx-io/gmx-interface/pull/370#discussion_r1071696707, is there are issue i've described?

Reply:
> and for this comment [#370 (comment)](https://github.com/gmx-io/gmx-interface/pull/370#discussion_r1071696707), is there are issue i've described?

Because we are using the function `setSymbol` from `tvWidgetRef.current` and it can be undefined I am using that so we can remove the error `'tvWidgetRef.current' is possibly 'null'`. 

There are two things `chartReady` that is stored in react state and `tvWidget` which is stored in ref. As of now I am using ref and I don't know why moving from ref for tvWidget to state doesn't work as with the ref (https://github.com/gmx-io/gmx-interface/pull/370#discussion_r1080861097). I am using chartReady to re-render the component on change of tvWidget. Once I find the answer for the above I will remove chartReady.

Reply:
not sure if i explained it correctly.

here `useEffect` hook uses `tvWidgetRef.current?.activeChart?.().symbol()` inside. however there is no `tvWidgetRef` inside hook dependencies. 

is it possible that `tvWidgetRef.current?.activeChart?.().symbol()` value changes, but hook would not call callback because dependencies didn't change?

for example in this case alert won't be called, right?:
```
const [a, setA] = useState(1);
const [b, setB] = useState(2);
setB(1);
useEffect(() => {
  if (a == b) { alert("equal"); }
}, [a]);
```